### PR TITLE
Store blob archives more efficient

### PR DESCRIPTION
**Pull request by tsteur**
refs #7373 

Please use this pull request only for implementation details on this pull request. If you do have objections or better ideas re the whole idea please comment on #7373

### CODE REVIEW

**Comment by mattab** [comment]
I think we should wait 2.13.0 for this change since it's hard to recover from it. Moving milestones.

File: core/Archive.php [hunk 1]
```diff
@@ -196,7 +197,7 @@ protected function __construct(Parameters $params, $forceIndexedBySite = false,
      *                             or date range (ie, 'YYYY-MM-DD,YYYY-MM-DD').
      * @param bool|false|string $segment Segment definition or false if no segment should be used. {@link Piwik\Segment}
      * @param bool|false|string $_restrictSitesToLogin Used only when running as a scheduled task.
-     * @return Archive
+     * @return static
```
Comment (diosmosis) [inline core/Archive.php]: Is there a reason for this other than for tests? If so, it might be good to explain the reasoning/use-cases for deriving from Archive in the PHP docs for the class.

If not, it might be cleaner to abstract the ::build/::factory methods to a new ArchiveFactory class that can be overridden via DI.

File: core/Archive.php [hunk 2]
```diff
@@ -319,6 +320,8 @@ public function getNumeric($names)
      *                                to the subtable ID to return. If set to 'all', all subtables
      *                                of each requested report are returned.
      * @return array An array of appropriately indexed blob data.
+     *
+     * @deprecated since Piwik 2.12. Use one of the getDatable* methods instead.
      */
```
Comment (diosmosis) [inline core/Archive.php]: Is there a reason for deprecating `getBlob()`. I agree it doesn't really serve a use nor is it a good idea to be exposed as an `@api` method since it's not backend agnostic, but perhaps it would be better to deprecate it in a separate pull request. Since it's not really related to this one.

File: tests/PHPUnit/Integration/ArchiveTest.php [hunk 3]
```diff
@@ -101,6 +114,125 @@ public function test_ArchiveIsLaunched_WhenForceOnBrowserRequest_IsFalse_AndArch
...
+        ));
+        $this->createArchiveBlobEntry('2013-01-03', array(
+            'Actions_Actionsurl' => array(0 => 'test03', 'subtables' => serialize(array(1 => 'subtable1', 2 => 'subtable2', 5 => 'subtable5'))),
+            'Actions_Actions' => array(0 => 'actions_03', 1 => 'actionsTest1', 2 => 'actionsTest2'))
+        );
+        $this->createArchiveBlobEntry('2013-01-04', array('Actions_Actionsurl' => array(0 => 'test04', 5 => 'subtable45', 6 => 'subtable6')));
+        $this->createArchiveBlobEntry('2013-01-06', array('Actions_Actionsurl' => array(0 => 'test06', 'subtables' => serialize(array()))));
+    }
+
+    private function assertArchiveBlob(PiwikArchive\DataCollection $dataCollection, $date, $expectedBlob)
```
Comment (diosmosis) [inline tests/PHPUnit/Integration/ArchiveTest.php]: It might increase clarity to append a verb to this method so we know what it does, eg, `assertArchiveBlobEquals` or `assertArchiveBlobHas`.

File: core/ArchiveProcessor.php [hunk 4]
```diff
@@ -327,7 +327,18 @@ public function insertNumericRecord($name, $value)
      */
     public function insertBlobRecord($name, $values)
     {
-        $this->archiveWriter->insertBlobRecord($name, $values);
+        $newInsert = array();
+
+        if (!empty($values[0])) {
```
Comment (diosmosis) [inline core/ArchiveProcessor.php]: I think it's not expected for `empty($values[0]) == true`, right? Would a log statement be potentially helpful in this case?

**Comment by diosmosis** [comment]
:+1: in general, as long as the increase in archiving time can be explained away and/or fixed.

**Comment by tsteur** [comment]
I will close pull request for now and issue a new one once finished
